cluster.go

```diff
@@ -169,62 +270,489 @@ func NewCluster() *Cluster {
+	return c.nodeByURI(c.Coordinator)
+}
+
+// IsCoordinator is true if this node is the coordinator.
+func (c *Cluster) IsCoordinator() bool {
+	return c.Static || c.Coordinator == c.Node.URI
+}
+
+// SetCoordinator updates the Coordinator to n.
+// Returns true if the Coordinator changed.
+func (c *Cluster) SetCoordinator(n *Node) bool {
+	// Get new node.
+	newNode := c.nodeByID(n.ID)
+	if newNode == nil {
+		return false
+	}
+
+	if c.Coordinator != newNode.URI {
+		c.Coordinator = newNode.URI
+		return true
 	}
+	return false
 }
 
-// NodeSetHosts returns the list of host strings for NodeSet members.
-func (c *Cluster) NodeSetHosts() []string {
-	if c.NodeSet == nil {
-		return []string{}
+// AddNode adds a node to the Cluster and updates and saves the
+// new topology.
+func (c *Cluster) AddNode(node *Node) error {
+	c.logger().Printf("add node %s to cluster on %s", node, c.Node)
+
+	// add to cluster
+	if !c.addNodeBasicSorted(node) {
+		return nil
+	}
+
+	// add to topology
+	if c.Topology == nil {
+		return fmt.Errorf("Cluster.Topology is nil")
 	}
-	a := make([]string, 0, len(c.NodeSet.Nodes()))
-	for _, m := range c.NodeSet.Nodes() {
-		a = append(a, m.Host)
+	if !c.Topology.AddID(node.ID) {
+		return nil
 	}
-	return a
+
+	// save topology
+	return c.saveTopology()
 }
 
-// NodeStates returns a map of nodes in the cluster with each node's state (UP/DOWN) as the value.
-func (c *Cluster) NodeStates() map[string]string {
-	h := make(map[string]string)
-	for _, n := range c.Nodes {
-		h[n.Host] = NodeStateDown
+// RemoveNode removes a node from the Cluster and updates and saves the
+// new topology.
+func (c *Cluster) RemoveNode(node *Node) error {
+	// remove from cluster
+	if !c.removeNodeBasicSorted(node) {
+		return nil
+	}
+
+	// remove from topology
+	if c.Topology == nil {
+		return fmt.Errorf("Cluster.Topology is nil")
 	}
-	// we are assuming that NodeSetHosts is a subset of c.Nodes
-	for _, m := range c.NodeSetHosts() {
-		if _, ok := h[m]; ok {
-			h[m] = NodeStateUp
+	if !c.Topology.RemoveID(node.ID) {
+		return nil
+	}
+
+	// save topology
+	return c.saveTopology()
+}
+
+// NodeIDs returns the list of IDs in the cluster.
+func (c *Cluster) NodeIDs() []string {
+	return Nodes(c.Nodes).IDs()
+}
+
+func (c *Cluster) setID(id string) {
+	// Don't overwrite ClusterID.
+	if c.ID != "" {
+		return
+	}
+	c.ID = id
+
+	// Make sure the Topology is updated.
+	c.Topology.ClusterID = c.ID
+}
+
+func (c *Cluster) State() string {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	return c.state
+}
+
+func (c *Cluster) SetState(state string) {
+	c.mu.Lock()
+	c.setState(state)
+	c.mu.Unlock()
+}
+
+func (c *Cluster) setState(state string) {
+	// Ignore cases where the state hasn't changed.
+	if state == c.state {
+		return
+	}
+
+	c.logger().Printf("change cluster state from %s to %s on %s", c.state, state, c.Node.ID)
+
+	var doCleanup bool
+
+	switch state {
+	case ClusterStateResizing:
+		c.prefect.SetRestricted()
+	case ClusterStateNormal:
```

Comment:
I think we should have a default case in this switch statement that panics

Reply:
addressed in 666fe22e